src/FilesystemSyncTaskHelper.php

```diff
@@ -0,0 +1,349 @@
+        /** @var FlysystemAssetStore $flysystemAssetStore */
+        $this->flysystemAssetStore = $assetStore;
+        /** @var PublicAssetAdapter $adapter */
+        $this->publicAssetAdapter = $this->flysystemAssetStore->getPublicFilesystem()->getAdapter();
+    }
+
+    private function deleteBroken()
+    {
+        if ($this->delete_broken) {
+            $this->deleteFilesWithMissingAssets();
+        }
+
+        DB::alteration_message(
+            $this->results['filesRemovedFromDb'] . ' files removed from the database (because they are missing files on the filesystem)',
+            "info"
+        );
+    }
+
+    /**
+     * @throws ValidationException
+     */
+    private function syncFiles()
+    {
+        foreach ($this->publicAssetAdapter->getFileGenerator($this->path, function ($file) {
+            return !$this->skipFile($file);
+        }) as $file) {
+            if ($this->canSyncFileToDatabase($file)) {
+                $this->syncFileToDatabase($file);
+            }
+
+            if ($file['type'] == 'dir') {
+                $realPath = $this->publicAssetAdapter->relativeToRealPath($file['path']);
+                $this->dirsToRemove[]=$realPath;
+            }
+        }
+
+        DB::alteration_message(
+            $this->results['filesSyncedWithDb'] . ' files on the filesystem synced with the database',
+            "info"
+        );
+        DB::alteration_message(
+            $this->results['dirsSyncedWithDb'] . ' directories on the filesystem synced with the database',
+            "info"
+        );
+        DB::alteration_message(
+            $this->results['filesRemovedFromFilesystem'] . ' hashed files removed from the filesystem (because they are missing matching database entries)',
+            "info"
+        );
+        DB::alteration_message(
+            $this->results['filesSkippedFromFilesystem'] . ' files skipped from the filesystem as they already have database entries',
+            "info"
+        );
+    }
+
+    private function removeDirectories()
+    {
+        foreach (array_reverse($this->dirsToRemove) as $dir) {
+            if (count(glob("$dir/*")) === 0) {
+                rmdir($dir);
+            }
+        }
+    }
+
+    /**
+     * Whether the provided file should be skipped for syncing or not.
+     * @param array $file
+     * @return bool
+     */
+    private function skipFile($file)
+    {
+        // skip files starting with:
+        $skipsStartingWith = [
+            '_',
+            '.'
+        ];
+        foreach ($skipsStartingWith as $skipStartingWith) {
+            if (substr(basename($file['path']), 0, strlen($skipStartingWith)) === $skipStartingWith) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Whether the provided file should be synced with the database or not.
+     * @param array $fileTuple
+     * @return bool
+     */
+    private function canSyncFileToDatabase(array $fileTuple)
```

Comment:
Big chunks of this method might need to be rewritten, once I'm done with my permalink jobs.

I'll leave some comments on thing, but maybe don't action them right away.